### ISSUE TITLE
Allow type parameters on multi effects

### ISSUE THREAD

**Issue by b-studios** [effekt-lang/effekt#15]
Currently only effect operations can take type parameters:

```
effect Foo {
  def op[A](...): ...
}
```

To express a state effect with operations get and put, we would like to also support type parameters on effects
```
effect State[S] { ... }
```

**Comment by b-studios** [comment]
This requires implementing unification of effect types in trait `Substitution`. Two effects should only be considered equal, if they are applied to the same type parameters.

**Comment by b-studios** [comment]
Addressed in #54
